---
title: "Candidate Cancer Gene Database"
bibliography: refs/ccgd_paper.bib
nocite: '@*'
output:
    html_document:
        includes:
            in_header: "styles/favicon.html"
            after_body: "styles/footer.html"
        css: styles/styles.css
        theme: readable
---

The Candidate Cancer Gene Database (CCGD) was developed to disseminate the results of transposon-based forward genetic screens in mice that identify candidate cancer genes. The purpose of the database is to allow cancer researchers to quickly determine whether or not a gene, or list of genes, has been identified as a potential cancer driver in a forward genetic screen in mice.

***

### Background

The work is a product of the [Starr Lab](http://www.med.umn.edu/starrlab/) at the [University of Minnesota](http://www.umn.edu/). The Starr Lab is affiliated with the [Center for Genome Engineering](http://www.cge.umn.edu/), the [Masonic Cancer Center](http://www.cancer.umn.edu/), the [Department of Obstetrics, Gynecology and Women's Health](http://www.obgyn.umn.edu/), and the [Department of Genetics, Cell Biology, and Development](http://www.cbs.umn.edu/gcd/).

##### Contributions

The idea for creating this database originated from Tim Starr, Vincent Keng, and David Largaespada at UMN. Erik Nyre and Ken Abbott developed the original database, with help from Juan Abrahante. A complete website and backend overhaul was conducted by Christopher Tastad in 2019.

* Ken Abbott
* [Erik Nyre](https://www.dorsey.com/people/n/nyre-erik)
* [Juan Abrahante](https://research.umn.edu/units/umii/about-us/staff-directory/juan-e-abrahante-phd)
* [Christopher Tastad](https://github.com/ctastad)

* [Tim Starr](https://med.umn.edu/bio/obgyn-faculty/tim-starr)
* [David Largaespada](https://www.pediatrics.umn.edu/bio/largaespada-lab-members/david-largaespada)
* [Vincent Keng](https://www.polyu.edu.hk/abct/en/staff/academic_staff/index.php?id=wvkeng)


The database is hosted by the Office of Information Technology at UMN. The database obtains regular updates from NCBI Gene and NCBI HomoloGene. The database also obtains regular updates from Sanger CGC and Sanger COSMIC. Administrators add information from NCBI PubMed. When necessary, administrators convert genome addresses to the current mouse genome build using UCSC LiftOver.

***

##### Updates

The most recent table build was completed on:

<object type="text/html"
        height=40px
        data="build_date.txt"></object>

***

##### Citation

Please use the following when citing the CCGD:

* Click <a href="refs/ccgd_paper.bib" download>here</a> for the citation in bibtex format.

